include/sundials/sundials_adjointcheckpointscheme.h

```diff
@@ -0,0 +1,152 @@
+                                               SUNAdjointCheckpointScheme*);
+
+/*
+ * Base class methods
+ */
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetNeedsSavingFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeNeedsSavingFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetNeedsDeletingFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeNeedsDeletingFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetInsertVectorFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeInsertVectorFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetLoadVectorFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeLoadVectorFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetRemoveVectorFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeRemoveVectorFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetDestroyFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeDestroyFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetEnableDenseFn(
+  SUNAdjointCheckpointScheme, SUNAdjointCheckpointSchemeEnableDenseFn);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_SetContent(SUNAdjointCheckpointScheme,
+                                                 void* content);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_GetContent(SUNAdjointCheckpointScheme,
+                                                 void** content);
+
+/*
+ * Virtual (overridable) base class methods
+ */
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_NeedsSaving(SUNAdjointCheckpointScheme,
+                                                  suncountertype step_num,
+                                                  suncountertype stage_num,
+                                                  sunrealtype t,
+                                                  sunbooleantype* yes_or_no);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_NeedsDeleting(SUNAdjointCheckpointScheme,
+                                                    suncountertype step_num,
+                                                    suncountertype stage_num,
+                                                    sunrealtype t,
+                                                    sunbooleantype* yes_or_no);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_InsertVector(SUNAdjointCheckpointScheme,
+                                                   suncountertype step_num,
+                                                   suncountertype stage_num,
+                                                   sunrealtype t, N_Vector state);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_LoadVector(
+  SUNAdjointCheckpointScheme, suncountertype step_num, suncountertype stage_num,
+  sunbooleantype peek, N_Vector* out, sunrealtype* tout);
+
+SUNDIALS_EXPORT
+SUNErrCode SUNAdjointCheckpointScheme_RemoveVector(SUNAdjointCheckpointScheme,
```

Comment:
When is this used? It doesn't seem to be called in the ARKODE branch

Reply:
I intended to use it in the ARKODE branch, but didnt. However, we should use it where `LoadVector` was being used to throw away a checkpoint.

Reply:
I think for now we can remove `RemoveVector` and `NeedsDeleting`. I think will be needed to implement more complex checkpointing, but for the fixed checkpointing what we have now is more efficient.

Reply:
Sounds good, I figured that as the case. Related, should the `NeedsDeleting` and `RemoveVector` names align i.e., either change `RemoveVector` to `DeleteVector` or `NeedsDeleting` to `NeedsRemoving`?

Reply:
Looks like we were commenting at the same time. Ignore my last comment if they're being removed.